### PR TITLE
Finish `realm.delete`

### PR DESCRIPTION
**Pull request by takameyer**
Add support for delete results and lists
Added tests for delete list of primitives

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 📝 Update `COMPATIBILITY.md`
* [x] 🚦 Tests
* [ ] 🔀 Executed flexible sync tests locally if modifying flexible sync
* [ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated

### CODE REVIEW

File: packages/realm-app-importer/src/DeleteAll.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+import fetch from "node-fetch";
+
+const getApps = async (headers: any) => {
```
Comment (kraenhansen) [inline packages/realm-app-importer/src/DeleteAll.ts:20]: Did you mean to add this? If yes, I suggest moving this to a different PR.

File: packages/realm/src/Realm.ts [hunk 2]
```diff
@@ -511,16 +511,19 @@ export class Realm {
...
-    } else if (Array.isArray(subject) || Symbol.iterator in subject) {
+    } else if (Array.isArray(subject) || subject instanceof Results) {
       // TODO: Optimize this to not get the helper on every iteration
       for (const object of subject) {
         assert.instanceOf(object, RealmObject);
         const { objectSchema } = this.classes.getHelpers(object);
         const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
         table.removeObject(object[INTERNAL].key);
       }
+    } else if (subject instanceof List) {
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:522]: I would move this branch up before the `Array.isArray`.

File: packages/realm-app-importer/src/DeleteAll.ts (hunk 1 above)
Comment (takameyer) [reply packages/realm-app-importer/src/DeleteAll.ts]: Nope

File: packages/realm/src/Realm.ts [hunk 3]
```diff
@@ -511,16 +511,19 @@ export class Realm {
       );
       const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
       table.removeObject(obj.key);
-    } else if (Array.isArray(subject) || Symbol.iterator in subject) {
+    } else if (Array.isArray(subject) || subject instanceof Results) {
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts]: I would add another branch above conditional on `subject instanceof Results` calling the `subject.clear()` (which needs to be added to the packages/bindgen/spec.yml first)

```suggestion
    } else if (Array.isArray(subject) || Symbol.iterator in subject) {
```

File: packages/realm/src/Realm.ts [hunk 4]
```diff
@@ -511,16 +511,24 @@ export class Realm {
       );
       const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
       table.removeObject(obj.key);
+    } else if (subject instanceof List) {
+      const internalList = subject.internal;
+      internalList.removeAll();
+    } else if (subject instanceof Results) {
+      const internalResult = subject.internal;
+      internalResult.clear();
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:519]: Little need for a local variable.
```suggestion
      subject.internal.clear();
```

File: packages/realm/src/Realm.ts [hunk 5]
```diff
@@ -511,16 +511,24 @@ export class Realm {
       );
       const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
       table.removeObject(obj.key);
+    } else if (subject instanceof List) {
+      const internalList = subject.internal;
+      internalList.removeAll();
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:516]: Little need for a local variable and [the existing SDK called deleteAll](https://github.com/realm/realm-js/blob/master/src/js_realm.hpp#L1408):

```suggestion
      subject.internal.deleteAll();
```

File: packages/realm/src/Realm.ts [hunk 6]
```diff
@@ -511,16 +511,24 @@ export class Realm {
...
+      const internalResult = subject.internal;
+      internalResult.clear();
     } else if (Array.isArray(subject) || Symbol.iterator in subject) {
-      // TODO: Optimize this to not get the helper on every iteration
-      for (const object of subject) {
-        assert.instanceOf(object, RealmObject);
-        const { objectSchema } = this.classes.getHelpers(object);
-        const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
-        table.removeObject(object[INTERNAL].key);
+      if (subject.length > 0) {
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:521]: Why is this check needed?

File: packages/realm/src/Realm.ts [hunk 7]
```diff
@@ -511,16 +511,24 @@ export class Realm {
...
+        // Get the objectSchema from the first element, and assume that all elements are of the same object type
+        const { objectSchema } = this.classes.getHelpers(subject[0]);
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:523]: This needs to happen after the assert that `object` is instance of `RealmObject`.

I actually think it would be better to leave it as it was and simply remove the TODO. It will be a lot less hot code now that you've added native support for `List` and `Result` and getting the helper on every object makes the API more useful as you can call it with an array of different objects.

File: packages/realm/src/Realm.ts (hunk 3 above)
Comment (takameyer) [reply packages/realm/src/Realm.ts]: This is done

File: packages/realm/src/Realm.ts (hunk 5 above)
Comment (takameyer) [reply packages/realm/src/Realm.ts]: Done.  Also did the same for Results

File: packages/realm/src/Realm.ts (hunk 7 above)
Comment (takameyer) [reply packages/realm/src/Realm.ts:521]: I could also assert `object` at that level as well.

Comment (takameyer) [reply packages/realm/src/Realm.ts:521]:
```typescript
        assert.instanceOf(subject[0], RealmObject);
        const { objectSchema } = this.classes.getHelpers(subject[0]);
        for (const object of subject) {
          const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
          table.removeObject(object[INTERNAL].key);
        }
```

Comment (takameyer) [reply packages/realm/src/Realm.ts:521]: But if for some reason, one was to just wildly put a bunch of realm objects of a different type into an array and try and delete it, this might cause some issues.  Maybe just putting it back the way it was is fine

File: packages/realm/src/Realm.ts [hunk 8]
```diff
@@ -511,16 +511,21 @@ export class Realm {
...
+    } else if (subject instanceof Results) {
+      subject.internal.clear();
     } else if (Array.isArray(subject) || Symbol.iterator in subject) {
-      // TODO: Optimize this to not get the helper on every iteration
-      for (const object of subject) {
-        assert.instanceOf(object, RealmObject);
-        const { objectSchema } = this.classes.getHelpers(object);
-        const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
-        table.removeObject(object[INTERNAL].key);
+      if (subject.length > 0) {
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:519]: No need for this check now.

```suggestion
```

**Review by kraenhansen — APPROVED** [review]
